pentaho-webpackage/extenders/http-extender/src/main/java/org/pentaho/webpackage/extender/http/impl/PentahoWebPackageServiceImpl.java

```diff
@@ -68,12 +70,59 @@ public void removeBundle( Bundle bundle ) {
       return;
     }
 
-    PentahoWebPackageBundle pwpc = this.pentahoWebPackageBundles.remove( bundle.getBundleId() );
+    PentahoWebPackageBundleImpl pwpc = this.pentahoWebPackageBundles.remove( bundle.getBundleId() );
     if ( pwpc != null ) {
       pwpc.destroy();
     }
   }
 
+  @Override
+  public PentahoWebPackageResource resolve( String moduleId ) {
+    int index = 0;
+
+    boolean moduleIdIncludesOrganization = moduleId.indexOf( '@' ) == 0;
+    if ( moduleIdIncludesOrganization ) {
+      index = moduleId.indexOf( '/', 1 ) + 1;
+    }
+
+    index = moduleId.indexOf( '/', index );
+
+    String baseModuleId = moduleId.substring( 0, index );
+
+    int versionSeparatorIndex = baseModuleId.lastIndexOf( '_' );
+
+    if ( versionSeparatorIndex == -1 ) {
+      return null;
+    }
+
+    String packageName = baseModuleId.substring( 0, versionSeparatorIndex );
+    String packageVersion = baseModuleId.substring( versionSeparatorIndex + 1 );
+
+    PentahoWebPackage webPackage = findWebPackage( packageName, packageVersion );
+    if ( webPackage == null ) {
+      return null;
+    }
+
+    String resourcePath = webPackage.getResourceRootPath() + "/" + moduleId.substring( index + 1 );
+    ClassLoader classLoader = webPackage.getBundle().adapt( BundleWiring.class ).getClassLoader();
+
+    return new PentahoWebPackageResourceImpl( resourcePath, classLoader );
+  }
+
+  PentahoWebPackage findWebPackage( String name, String version ) {
+    synchronized ( this.pentahoWebPackageBundles ) {
```

Comment:
Some kind of ReaderWriter lock, here and `addBundle`, would be much more efficient, given that we're almost always reading.

Reply:
@dcleao `addBundle` is only called when there is a Bundle changed event and that doesn't happen that often.
So I don't think this will impact performance that much.